apluslms_roman/validator.py

```diff
@@ -0,0 +1,85 @@
+import yaml
+from jsonschema import validate
+import re
+import os
+#from operator import itemgetter
+import logging
+
+logger = logging.getLogger(__name__)
+
+#mystinen lista schema directoryja - course.yamlissa jossain kentässä, joka sijaitsee käyttäjän paketin juuressa
+def get_folders(config):
+    folders = config.__config__.get("custom_schema_locations", [])
+    dir_, _ = os.path.split(__file__)
+    folders.append(os.path.join(dir_, "schemas"))
+    return folders
+
+
+#idea = { schema_name: {(major, minor): location}}
+schema_index = {}
+
+class Validator:
+    def __init__(self, config):
+        logger.info(" Creating validator.")
+        self.folders = get_folders(config)
+        logger.info(" Validator ready.")
+
+    def validate(self, filename, schema, major=1, minor=float('inf')):
+        logger.info(" Preparing to validate %s...", filename)
+        logger.info(" Locating all valid schemas...")
+        self.find_all_matches(schema)
+        logger.info(" Locating matching version v%d...", major)
+        serial, ext = self.find_newest(schema, major, minor)
+        schema_fullname = "{}_v{}_{}.{}".format(schema, serial[0], serial[1], ext[1])
+        schema_path = os.path.join(ext[0], schema_fullname)
+        logger.info(" Starting validation of %s with %s", filename, schema_fullname)
+        result = self.assert_valid(schema_path, filename)
+        logger.info(" Validation done", filename)
+        return result
+
+    def find_all_matches(self, schema):
+        prog = re.compile("{}_v(\d+)_(\d+).(yaml|yml|json)".format(schema))
+        logger.debug(" Looking for schema-------------%s", schema)
+        for folder in self.folders:
+            logger.debug(" Folder-------------%s",  folder)
+            for ff in os.listdir(folder):
+                logger.debug(" File-------------%s",  ff)
+                if prog.match(ff) is not None:
+                    found = prog.match(ff).groups()
+                    schema_index.setdefault(schema, {})
```

Comment:
`foo = schema_index.setdefault(...)` store result in var and use it below in place of `schema_index[schema]`.